net/net.go

```diff
@@ -848,110 +860,183 @@ func (n *net) PutRecord(ctx context.Context, id thread.ID, lid peer.ID, rec core
 }
 
 // putRecord adds an existing record. This method is thread-safe.
-func (n *net) putRecord(ctx context.Context, id thread.ID, lid peer.ID, rec core.Record) error {
-	tsph := n.getThreadSemaphore(id)
-	tsph <- struct{}{}
-	defer func() { <-tsph }()
-	return n.putRecordUnsafe(ctx, id, lid, rec)
-}
+func (n *net) putRecord(ctx context.Context, tid thread.ID, lid peer.ID, rec core.Record) error {
+	unknown, head, err := n.loadUnknownRecords(ctx, tid, lid, rec)
+	if err != nil {
+		return fmt.Errorf("loading records failed: %w", err)
+	} else if len(unknown) == 0 {
+		return nil
+	}
 
-// putRecordUnsafe adds an existing record. See PutOption for more. This method
-// *should be thread-guarded*.
-func (n *net) putRecordUnsafe(ctx context.Context, id thread.ID, lid peer.ID, rec core.Record) error {
-	var unknownRecords []core.Record
-	c := rec.Cid()
-	for c.Defined() {
-		exist, err := n.bstore.Has(c)
-		if err != nil {
+	ts := n.semaphores.Get(threadSemaphore(tid))
+	ts.Acquire()
+
+	// check head again to detect if some other process concurrently have changed the log
+	if current, err := n.currentHead(tid, lid); err != nil {
+		ts.Release()
+		return fmt.Errorf("fetching head failed: %w", err)
+	} else if current != head {
+		ts.Release()
+		return n.putRecord(ctx, tid, lid, rec)
+	}
+
+	defer ts.Release()
+
+	connector, appConnected := n.getConnector(tid)
+	for _, record := range unknown {
+		if err := n.store.SetHead(tid, lid, record.Value().Cid()); err != nil {
+			return fmt.Errorf("setting log head failed: %w", err)
+		}
+
+		if appConnected {
+			if err := connector.HandleNetRecord(ctx, record); err != nil {
+				// Future improvement notes.
+				// If record handling fails there are two options available:
+				// 1. Just interrupt and return error (current behaviour). Log head remains moved and some events
+				//    from the record possibly won't reach reducers/listeners or even get dispatched.
+				// 2. Rollback log head to the previous record. In this case record handling will be retried until
+				//    success, but reducers must guarantee its idempotence and there is a chance of getting stuck
+				//    with bad event and not making any progress at all.
```

Comment:
Good notes! The original thinking here is inline with your warning in (2)... if it failed the first time, chances are it will fail again.